onedal/primitives/tree_visitor.cpp

```diff
@@ -0,0 +1,344 @@
+    auto value_ar_strides = py::array::StridesContainer(
+        { this->class_count * sizeof(double), this->class_count * sizeof(double), sizeof(double) });
+
+    skl_tree_node* node_ar_ptr = new skl_tree_node[this->node_count];
+    double* value_ar_ptr =
+        new double[this->node_count * 1 *
+                   this->class_count](); // oneDAL only supports scalar responses for now
+
+    this->node_ar = py::array_t<skl_tree_node>(node_ar_shape, node_ar_strides, node_ar_ptr);
+    this->value_ar = py::array_t<double>(value_ar_shape, value_ar_strides, value_ar_ptr);
+}
+
+template <typename Task>
+bool to_sklearn_tree_object_visitor<Task>::call(const df::split_node_info<Task>& info) {
+    py::buffer_info node_ar_buf = this->node_ar.request();
+
+    skl_tree_node* node_ar_ptr = static_cast<skl_tree_node*>(node_ar_buf.ptr);
+
+    if (info.get_level() > 0) {
+        // has parents
+        Py_ssize_t parent = parents[info.get_level() - 1];
+        if (node_ar_ptr[parent].left_child > 0) {
+            assert(node_ar_ptr[node_id].right_child < 0);
+            node_ar_ptr[parent].right_child = node_id;
+        }
+        else {
+            node_ar_ptr[parent].left_child = node_id;
+        }
+    }
+    parents[info.get_level()] = node_id;
+    node_ar_ptr[node_id].feature = info.get_feature_index();
+    node_ar_ptr[node_id].threshold = info.get_feature_value();
+    node_ar_ptr[node_id].impurity = info.get_impurity();
+    node_ar_ptr[node_id].n_node_samples = info.get_sample_count();
+    node_ar_ptr[node_id].weighted_n_node_samples = info.get_sample_count();
+
+    // wrap-up
+    ++node_id;
+    return true;
+}
+
+// stuff that is done for all leaf node types
+template <typename Task>
+void to_sklearn_tree_object_visitor<Task>::_onLeafNode(const df::leaf_node_info<Task>& info) {
+    py::buffer_info node_ar_buf = this->node_ar.request();
+
+    skl_tree_node* node_ar_ptr = static_cast<skl_tree_node*>(node_ar_buf.ptr);
+
+    if (info.get_level()) {
+        Py_ssize_t parent = parents[info.get_level() - 1];
+        if (node_ar_ptr[parent].left_child > 0) {
+            assert(node_ar_ptr[node_id].right_child < 0);
+            node_ar_ptr[parent].right_child = node_id;
+        }
+        else {
+            node_ar_ptr[parent].left_child = node_id;
+        }
+    }
+
+    node_ar_ptr[node_id].impurity = info.get_impurity();
+    node_ar_ptr[node_id].n_node_samples = info.get_sample_count();
+    node_ar_ptr[node_id].weighted_n_node_samples = info.get_sample_count();
+}
+
+template <>
+bool to_sklearn_tree_object_visitor<df::task::regression>::call(
+    const df::leaf_node_info<df::task::regression>& info) {
+    _onLeafNode(info);
+    OVERFLOW_CHECK_BY_MULTIPLICATION(int, node_id, class_count);
+
+    py::buffer_info value_ar_buf = this->value_ar.request();
+    double* value_ar_ptr = static_cast<double*>(value_ar_buf.ptr);
+
+    value_ar_ptr[node_id * 1 * this->class_count] = info.get_response();
```

Comment:
I think there also should be a check that the index `node_id * 1 * this->class_count` is not greater than the `value_ar_ptr` buffer size.